image-hd.c

```diff
@@ -166,6 +199,160 @@ static int hdimage_insert_ebr(struct image *image, struct partition *part)
+	case 'L': return "0fc63daf-8483-4772-8e79-3d69d8477de4";
+	case 'S': return "0657fd6d-a4ab-43c4-84e5-0933c84b4f4f";
+	case 'H': return "933ac7e1-2eb4-4f13-b844-0e14e2aef915";
+	case 'U': return "c12a7328-f81f-11d2-ba4b-00a0c93ec93b";
+	case 'R': return "a19d880f-05fc-4d3b-a006-743f0f84911e";
+	case 'V': return "e6d6d379-f507-44c2-a23c-238f2a3df928";
+	case 'F': return "ebd0a0a2-b9e5-4433-87c0-68b6b72699c7";
+	}
+	return str;
+}
+
+
+static int hdimage_insert_protective_mbr(struct image *image)
+{
+	struct partition mbr;
+	struct list_head mbr_list = LIST_HEAD_INIT(mbr_list);
+	int ret = 0;
+
+	image_info(image, "writing protective MBR\n");
+
+	memset(&mbr, 0, sizeof(struct partition));
+	mbr.offset = 512;
+	mbr.size = image->size - 512;
+	mbr.in_partition_table = 1;
+	mbr.partition_type = 0xee;
+	list_add_tail(&mbr.list, &mbr_list);
+	ret = hdimage_insert_mbr(image, &mbr_list);
+	if (ret) {
+		image_error(image,"failed to write protective MBR\n");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int hdimage_insert_gpt(struct image *image, struct list_head *partitions)
+{
+	struct hdimage *hd = image->handler_priv;
+	const char *outfile = imageoutfile(image);
+	struct gpt_header header;
+	struct gpt_partition_entry table[GPT_ENTRIES];
+	struct partition *part;
+	unsigned i, j;
+	int ret;
+
+	image_info(image, "writing GPT\n");
+
+	memset(&header, 0, sizeof(struct gpt_header));
+	memcpy(header.signature, "EFI PART", 8);
+	header.revision = htole32(GPT_REVISION_1_0);
+	header.header_size = htole32(sizeof(struct gpt_header));
+	header.current_lba = htole64(1);
+	header.backup_lba = htole64(image->size/512 - 1);
+	header.last_usable_lba = htole64(image->size/512 - 1 - GPT_SECTORS);
+	if (hd->disk_uuid) {
+		if (uuid_parse(hd->disk_uuid, header.disk_uuid) == -1) {
+			image_error(image, "invalid disk UUID: %s\n", hd->disk_uuid);
+			return -EINVAL;
+		}
+	}
+	else {
+		uuid_random(header.disk_uuid);
+	}
+	header.starting_lba = htole64(2);
+	header.number_entries = htole32(GPT_ENTRIES);
+	header.entry_size = htole32(sizeof(struct gpt_partition_entry));
+
+	i = 0;
+	memset(&table, 0, sizeof(table));
+	list_for_each_entry(part, partitions, list) {
+		if (header.first_usable_lba == 0)
+			header.first_usable_lba = htole64(part->offset / 512);
+
+		if (!part->in_partition_table)
+			continue;
+
+		if (strlen(part->partition_type_uuid) == 1)
+			part->partition_type_uuid =
+				gpt_partition_type_lookup(part->partition_type_uuid);
+		if (uuid_parse(part->partition_type_uuid, table[i].type_uuid) == -1) {
+			image_error(image,
+				    "part %s has invalid partition type UUID: %s\n",
+				    part->name, part->partition_type_uuid);
+			return -EINVAL;
+		}
+		if (part->partition_uuid) {
+			if (uuid_parse(part->partition_uuid, table[i].uuid) == -1) {
+				image_error(image,
+					    "part %s has invalid partition UUID: %s\n",
+					    part->name, part->partition_uuid);
+				return -EINVAL;
+			}
+		}
+		else {
+			uuid_random(table[i].uuid);
+		}
```

Comment:
Also in setup. Any parsing / validating of the config should be handled there.
Basically only I/O errors should cause generate to fail,